packages/twenty-server/src/engine/core-modules/user/services/user.service.ts

```diff
@@ -243,4 +267,161 @@ export class UserService extends TypeOrmQueryService<User> {
+        );
+
+      if (isPermissionsV2Enabled) {
+        const permissions =
+          await this.permissionsService.getUserWorkspacePermissionsV2({
+            userWorkspaceId: currentUserWorkspace.id,
+            workspaceId: workspace.id,
+          });
+
+        settingsPermissions = permissions.settingsPermissions;
+        objectPermissions = Object.entries(permissions.objectPermissions).map(
+          ([objectMetadataId, permissions]) => ({
+            objectMetadataId,
+            canReadObjectRecords: permissions.canRead,
+            canUpdateObjectRecords: permissions.canUpdate,
+            canSoftDeleteObjectRecords: permissions.canSoftDelete,
+            canDestroyObjectRecords: permissions.canDestroy,
+          }),
+        );
+        objectRecordsPermissions = permissions.objectRecordsPermissions;
+      } else {
+        const permissions =
+          await this.permissionsService.getUserWorkspacePermissions({
+            userWorkspaceId: currentUserWorkspace.id,
+            workspaceId: workspace.id,
+          });
+
+        settingsPermissions = permissions.settingsPermissions;
+        objectRecordsPermissions = permissions.objectRecordsPermissions;
+      }
+    }
+
+    const grantedSettingsPermissions: SettingPermissionType[] = (
+      Object.keys(settingsPermissions) as SettingPermissionType[]
+    ).filter((feature) => settingsPermissions[feature] === true);
+
+    const grantedObjectRecordsPermissions = (
+      Object.keys(objectRecordsPermissions) as PermissionsOnAllObjectRecords[]
+    ).filter((permission) => objectRecordsPermissions[permission] === true);
+
+    currentUserWorkspace.settingsPermissions = grantedSettingsPermissions;
+    currentUserWorkspace.objectRecordsPermissions =
+      grantedObjectRecordsPermissions;
+    currentUserWorkspace.objectPermissions = objectPermissions;
+
+    return currentUserWorkspace;
+  }
+
+  private async getUserAvailableWorkspaces(
+    user: User,
+  ): Promise<AvailableWorkspacesToJoin[]> {
+    const domain = getDomainNameByEmail(user.email);
+    const userWorkspaceIds = user.workspaces.map(
+      ({ workspaceId }) => workspaceId,
+    );
+
+    const [approvedDomainWorkspaces, invitedWorkspaces] = await Promise.all([
+      (
+        await this.approvedAccessDomainService.findValidatedApprovedAccessDomainWithWorkspaceByDomain(
+          domain,
+        )
+      ).map(({ workspace }) =>
+        this.mapWorkspaceToAvailableWorkspace(workspace),
+      ),
+      (
+        await this.workspaceInvitationService.findInvitationsByEmail(user.email)
+      ).map((appToken) =>
+        this.mapWorkspaceToAvailableWorkspace(appToken.workspace, appToken),
+      ),
+    ]);
+
+    return [...approvedDomainWorkspaces, ...invitedWorkspaces].filter(
+      ({ id }) => !userWorkspaceIds.includes(id),
+    );
+  }
+
+  private mapWorkspaceToAvailableWorkspace(
+    workspace: Workspace,
+    appToken?: AppToken,
+  ): AvailableWorkspacesToJoin {
```

Comment:
question: I feel that the AvailableWorkspacesToJoin is more like a PublicWorkspaceInfo